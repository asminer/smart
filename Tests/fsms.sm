
// $Id$

/* 
   Tests for fsm formalism and internal stuff
*/

#GraphDisplayStyle OUTGOING

// A small simple one to check the num_states and num_arcs functions

fsm d0 := {
  state a,b;
  init({a});
  arcs(
  	a:a,
	a:b,
	b:a
  );

  void show := { 
    print("d0 states:\n"); 
    show_states; 
    print("d0 ");
    show_arcs;
  };
};

d0.show;

//
// Check state errors
//

print("Some bad state name errors:\n");

fsm d1 := {
  state a,b,a;
  init({a});
  arcs(
  	a:c
  );
};

// 
// Warnings for duplicate arcs and such
//

fsm d2 := {
  state a,b,c;
  init({a, b, a});  // no error
  arcs(
    a:b,
    a:c,
    a:b
  );
  bigint m1 := num_states;
};

print("Warnings for duplicates:\n");
print(d2.m1, "\n");

//
// No classification for FSMs
//

fsm d3(int dummy) := {
  state a, 	// absorbing
        b,c,d,	// recurrent class
	e,f,	// recurrent class
	g,h,i,j;  // transients

  init ({i});

  arcs(
  	i : g,
	g : h,
	h : i,

	g : a,
	h : b,
	i : j, j : e,

	e : e, e : f,
	f : e,

	b : c, b : d,
	c : d,
	d : b
  );

  void show := { 
    print("d3 states:\n"); 
    show_states; 
    print("d3 ");
    show_arcs;
  };
};

# StateDisplayOrder LEXICAL
print("States should appear in definition order: a,b,c,d,e,f,g,h,i,j:\n");
d3(1).show;

# StateDisplayOrder NATURAL
print("States should still appear in definition order: a,b,c,d,e,f,g,h,i,j:\n");
d3(2).show;


//
// Simple array of states test
//

fsm d4(int N) := {
  for (int i in {1..N}) {
    state s[i];
  }
  init({s[1]});
  for (int i in {2..N}) {
    arcs(s[i-1] : s[i], s[i] : s[i-1]);
  }
  void show := { 
    show_states; 
    show_arcs;
  };
};

for (int n in {1..5}) {
  print("Birth death with ", n, " states:\n");
  d4(n).show;
}

//
// crazy stuff that should not be allowed
// 

fsm foo1(state a, state b) := {
  state c, d, e;
  init({b, c});
  arcs(
  	a:b, a:c,
	b:c, b:d,
	c:d, c:e,
	d:e,
	e:a
  );
  bigint m1 := num_states;
  stateset m2 := potential(in_state(a));
  stateset m3 := potential(in_state({a, b}));
};

fsm bar1 := {
  state x, y;
  init({x});
  bigint m1 := foo1(x, y).m1;
  stateset m2 := foo1(x, y).m2;
  stateset m3 := foo1(x, y).m3;
};

print("Should generate meaningful warnings:\n");
compute(bar1.m1);
print("This should be emptyset:\n");
print(bar1.m2, "\n");
print("This should be emptyset:\n");
print(bar1.m3, "\n");

print("Done with fsm tests.\n");
