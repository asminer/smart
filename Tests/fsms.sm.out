d0 states:
State 0: a
State 1: b
d0 Reachability graph:
From state 0:
	To state 0
	To state 1
From state 1:
	To state 0
Some bad state name errors:
ERROR in file fsms.sm near line 38:
	Duplicate identifier a within model
ERROR in file fsms.sm near line 42:
	Unknown identifier: c
ERROR in file fsms.sm near line 42:
	No match for arcs(error)
Warnings for duplicates:
WARNING in file fsms.sm near line 56:
	Ignoring duplicate arc from state a to b
	within model d2 built in file fsms.sm near line 61
3
States should appear in definition order: a,b,c,d,e,f,g,h,i,j:
d3 states:
State 0: a
State 1: b
State 2: c
State 3: d
State 4: e
State 5: f
State 6: g
State 7: h
State 8: i
State 9: j
d3 Reachability graph:
From state 0:
From state 1:
	To state 2
	To state 3
From state 2:
	To state 3
From state 3:
	To state 1
From state 4:
	To state 4
	To state 5
From state 5:
	To state 4
From state 6:
	To state 0
	To state 7
From state 7:
	To state 1
	To state 8
From state 8:
	To state 6
	To state 9
From state 9:
	To state 4
States should still appear in definition order: a,b,c,d,e,f,g,h,i,j:
d3 states:
State 0: a
State 1: b
State 2: c
State 3: d
State 4: e
State 5: f
State 6: g
State 7: h
State 8: i
State 9: j
d3 Reachability graph:
From state 0:
From state 1:
	To state 2
	To state 3
From state 2:
	To state 3
From state 3:
	To state 1
From state 4:
	To state 4
	To state 5
From state 5:
	To state 4
From state 6:
	To state 0
	To state 7
From state 7:
	To state 1
	To state 8
From state 8:
	To state 6
	To state 9
From state 9:
	To state 4
Birth death with 1 states:
State 0: s[1]
Reachability graph:
From state 0:
Birth death with 2 states:
State 0: s[1]
State 1: s[2]
Reachability graph:
From state 0:
	To state 1
From state 1:
	To state 0
Birth death with 3 states:
State 0: s[1]
State 1: s[2]
State 2: s[3]
Reachability graph:
From state 0:
	To state 1
From state 1:
	To state 0
	To state 2
From state 2:
	To state 1
Birth death with 4 states:
State 0: s[1]
State 1: s[2]
State 2: s[3]
State 3: s[4]
Reachability graph:
From state 0:
	To state 1
From state 1:
	To state 0
	To state 2
From state 2:
	To state 1
	To state 3
From state 3:
	To state 2
Birth death with 5 states:
State 0: s[1]
State 1: s[2]
State 2: s[3]
State 3: s[4]
State 4: s[5]
Reachability graph:
From state 0:
	To state 1
From state 1:
	To state 0
	To state 2
From state 2:
	To state 1
	To state 3
From state 3:
	To state 2
	To state 4
From state 4:
	To state 3
Should generate meaningful warnings:
WARNING in file fsms.sm near line 138:
	state y is from another model, ignoring as initial state
	within model foo1 built in file fsms.sm near line 154
WARNING in file fsms.sm near line 140:
	state x is from another model, ignoring arc
	within model foo1 built in file fsms.sm near line 154
WARNING in file fsms.sm near line 140:
	state x is from another model, ignoring arc
	within model foo1 built in file fsms.sm near line 154
WARNING in file fsms.sm near line 141:
	state y is from another model, ignoring arc
	within model foo1 built in file fsms.sm near line 154
WARNING in file fsms.sm near line 141:
	state y is from another model, ignoring arc
	within model foo1 built in file fsms.sm near line 154
WARNING in file fsms.sm near line 145:
	state x is from another model, ignoring arc
	within model foo1 built in file fsms.sm near line 154
This should be emptyset:
{}
This should be emptyset:
{}
Done with fsm tests.
